Link more assemblies (type granularity linking for real this time)

SteveSandersonMS opened the pull request:
This uses @marek-safar's technique for enabling type-granularity linking using the existing XML config capabilities.

It reduces the out-of-box standalone app size to around 1.93MB. This isn't the theoretical minimum we could achieve. We could get it down to about 1.85MB if we were willing to make more fine-grained choices about which assemblies from `Microsoft.AspNetCore.*` and `Microsoft.Extensions.*` require type-granularity linking and which don't. That's a risk because if we're wrong then it could fail at runtime if you're using some obscure API that we didn't anticipate. So, in this PR, I'm making a risk-averse choice of saying all those in `Microsoft.AspNetCore.*` and `Microsoft.Extensions.*` require type-granularity linking.

### Why there's a new MSBuild task that generates XML configs during build

It's necessary to generate the config for this dynamically during the build (and not just embed xml files into the assemblies) because:

 1. We need this type-granularity rule to apply to some assemblies that we don't own in this repo (e.g., things in `Microsoft.Extensions.*`, because they contains DI services, e.g., for options). So we'd have to put the rules for those in some other assembly.
 2. Sometimes the application might *not* reference those other assemblies
 3. The linker throws if you try to provide config for an assembly you're not referencing

As such it's not enough to have a fixed set of XML configs. We have to generate some config during the build based on which assemblies are in the transitive closure of your references.

SteveSandersonMS committed:
Make apps smaller via type-granularity linking

SteveSandersonMS commented:
I've realised the approach here isn't going to be sufficient. The problem is that we're defining things in two different places:

 1. We specify which assemblies should be linked in `Blazor.MonoRuntime.targets` using wildcard patters (e.g., include `Microsoft.AspNetCore.*`)
 2. We specify which assemblies should have type-level linking granularity in a set of separate `LinkerConfig.xml` files, because we don't know which dependencies will be included in any given app, and the linker throws if you specify config for an assembly that isn't referenced at all.

There is potential for inconsistency with this approach. Imagine there's some other assembly we never thought about before (`Microsoft.AspNetCore.SomethingNew`) which contains DI services. Now (1) will say it should be linked based on the wildcard match, but (2) doesn't know about it so won't configure it for type-granularity linking. Then at runtime it will fail because the DI service constructors will have been removed. So this XML-config based approach isn't as convenient as the command-line flags approach.

As a solution, I plan to remove all the `LinkerConfig.xml` files, and instead add a new build task that generates a single XML config file based on which items matched by (1) have some MSBuild metadata saying we want type-granularity linking. This will ensure we always have a consistent view, and unanticipated assemblies will be handled correctly.

SteveSandersonMS committed:
Generate the typegranularity linker config based on referenced assemblies

SteveSandersonMS committed:
Fix

pranavkm commented on src/Components/Blazor/Build/src/Tasks/GenerateTypeGranularityLinkingConfig.cs:
```diff
@@ -0,0 +1,61 @@
+        public override bool Execute()
+        {
+            using (var fileStream = File.Open(OutputPath, FileMode.Create))
+            using (var xmlWriter = XmlWriter.Create(fileStream, new XmlWriterSettings { Indent = true }))
```
Could this be changed to use `XElement` to be consistent with https://github.com/dotnet/aspnetcore/blob/master/src/Components/Blazor/Build/src/Tasks/BlazorCreateRootDescriptorFile.cs?

SteveSandersonMS replied:
Sure, done

pranavkm commented on src/Components/Blazor/Build/src/Tasks/GenerateTypeGranularityLinkingConfig.cs:
```diff
@@ -0,0 +1,61 @@
+
+            xmlWriter.WriteStartElement("type");
+            xmlWriter.WriteAttributeString("fullname", "*");
+            xmlWriter.WriteAttributeString("preserve", "all");
```
It might help to explain what setting `preserve=all` and `required=false` do.

SteveSandersonMS replied:
OK, done

pranavkm commented on src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -171,7 +172,10 @@
-      <_BlazorDependencyAssembly Include="@(_BlazorDependencyInput)" IsLinkable="$([System.String]::Copy('%(FileName)').StartsWith('System.'))" />
+      <_BlazorDependencyAssembly Include="@(_BlazorDependencyInput)" />
+      <_BlazorDependencyAssembly IsLinkable="true" Condition="$([System.String]::Copy('%(Filename)').StartsWith('System.'))" />
+      <_BlazorDependencyAssembly IsLinkable="true" TypeGranularity="true" Condition="$([System.String]::Copy('%(Filename)').StartsWith('Microsoft.AspNetCore.'))" />
```
Doesn't need to be as part of this PR, but we might want to think about having a way for the user to toggle the defaults.

SteveSandersonMS replied:
I agree. It would be nice if component libraries had an easy way to say "I'm safe to be linked". However it's out of scope for this PR.

pranavkm commented on src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -201,6 +205,14 @@
+    <PropertyGroup>
+      <_TypeGranularityLinkingConfig>$(BlazorIntermediateOutputPath)linker.typegranularityconfig.xml</_TypeGranularityLinkingConfig>
+    </PropertyGroup>
+    <GenerateTypeGranularityLinkingConfig Assemblies="@(_BlazorAssemblyToLink)" OutputPath="$(_TypeGranularityLinkingConfig)" />
```
It might be ever so slightly nicer to filter in MSBuild rather than in the task

pranavkm replied:
```suggestion
    <GenerateTypeGranularityLinkingConfig Assemblies="@(_BlazorAssemblyToLink->WithMetadataValue('TypeGranularity', 'true'))" OutputPath="$(_TypeGranularityLinkingConfig)" />
```

SteveSandersonMS replied:
Sure, will do. Previously I thought it was more flexible to pass more info into the task in case it wanted to do something with the other assemblies, but since this is all internal, we're free to change it in the future if there's ever a reason to do so.

Right now I agree it's more direct and obvious to pass a pre-filtered list to the task.

pranavkm commented on src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -201,6 +205,14 @@
+    <PropertyGroup>
+      <_TypeGranularityLinkingConfig>$(BlazorIntermediateOutputPath)linker.typegranularityconfig.xml</_TypeGranularityLinkingConfig>
+    </PropertyGroup>
+    <GenerateTypeGranularityLinkingConfig Assemblies="@(_BlazorAssemblyToLink)" OutputPath="$(_TypeGranularityLinkingConfig)" />
```
Invoking this target should be incremental. For most users, the contents of this file would almost never change. If you'd like to do less MSBuild, you can check this in and make an issue to make it so.

SteveSandersonMS replied:
This runs if and only if we're going to run the linker. Since the linker is orders of magnitude more expensive than this anyway, I didn't think it would achieve anything to make this bit incremental.

If you still think it's not right, please either file an issue so we track changing it in the future, or just let me know and I will do so. Thanks!

pranavkm reviewed (commented):
Seems pretty straight forward

rynowak commented:
@SteveSandersonMS - did you delete a comment? I saw a big comment from you in my notifications, but now it's gone.

SteveSandersonMS commented:
> did you delete a comment

Yes, previously this PR contained a completely different approach to solving this problem, and I wrote a comment saying "I realised this approach won't work, here's why, and what I will change". Then I changed the implementation completely based on what I had learned. Then I deleted that comment in an attempt to avoid confusion, and ironically created confusion by doing so.

SteveSandersonMS committed:
Pre-filter the assembly list when calling GenerateTypeGranularityLinkingConfig

Co-Authored-By: Pranav K <prkrishn@hotmail.com>

SteveSandersonMS committed:
CR feedback

SteveSandersonMS committed:
CR: Switch to use XElement

SteveSandersonMS committed:
Super minor tweak

rynowak commented:
> Then I changed the implementation completely based on what I had learned. Then I deleted that comment in an attempt to avoid confusion, and ironically created confusion by doing so.

Cool. It worked.

rynowak commented on src/Components/Blazor/Blazor/src/Hosting/WebAssemblyHost.cs:
```diff
@@ -19,6 +19,11 @@ internal class WebAssemblyHost : IWebAssemblyHost
+            // To ensure JS-invoked methods don't get linked out, have a reference to their enclosing types
+            GC.KeepAlive(typeof(EntrypointInvoker));
+            GC.KeepAlive(typeof(JSInteropMethods));
+            GC.KeepAlive(typeof(WebAssemblyEventDispatcher));
```
FYI there's an attribute called `[PreserveDependency]` that has this effect. I'm not sure about the status of it right now, so I'm not suggesting you change the code, but this kind of usage is getting thought about.

SteveSandersonMS replied:
I'm aware of it, but didn't want to use it because you have to specify the dependency as a string in a particular format. I preferred to declare the dependency in a strongly-typed way so it will survive future refactoring.

Also you have to create your own `PreserveDependencyAttribute` which would be OK if we were using it a lot in this assembly, but it feels like a bunch of extra noise for something handled more conveniently via these three lines.

pranavkm commented on src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets:
```diff
@@ -201,6 +205,14 @@
+    </PropertyGroup>
+    <GenerateTypeGranularityLinkingConfig Assemblies="@(_BlazorAssemblyToLink->WithMetadataValue('TypeGranularity', 'true'))" OutputPath="$(_TypeGranularityLinkingConfig)" />
+    <ItemGroup>
+      <BlazorLinkerDescriptor Include="$(_TypeGranularityLinkingConfig)" />
```
```suggestion
      <BlazorLinkerDescriptor Include="$(_TypeGranularityLinkingConfig)" />
      <FileWrites Include="$(_TypeGranularityLinkingConfig)" />
```

SteveSandersonMS committed:
Update src/Components/Blazor/Build/src/targets/Blazor.MonoRuntime.targets

Co-Authored-By: Pranav K <prkrishn@hotmail.com>
